BrainPortal/lib/boutiques_dir_maker.rb

```diff
@@ -0,0 +1,166 @@
+
+  # For input in the `BoutiquesDirMaker` section,
+  # create a directory based on the value BoutiquesDirMaker[input_id]
+  # that will contains a symlink to the orginal selected Userfile.
+  def setup #:nodoc:
+    return false unless super
+
+    descriptor = self.descriptor_for_setup
+    self.addlog(descriptor.file_revision_info.format("%f rev. %s %a %d"))
+
+    boutiques_dir_maker_list = descriptor.custom_module_info('BoutiquesDirMaker')
+    descriptor.file_inputs.each do |input|
+      userfile_id = invoke_params[input.id]
+      next if userfile_id.blank? # that happens when it's an optional file
+      userfile    = Userfile.find(userfile_id)
+      dirname     = boutiques_dir_maker_list[input.id]
+
+      # Most common situation
+      if ! input.list || ! userfile.is_a?(CbrainFileList)
+        make_available(userfile, dirname.present? ? "#{dirname}/#{userfile.name}" : userfile.name)
+        next
+      end
+
+      # In case the input is a list and is assigned a CbrainFileList
+      userfile.sync_to_cache
+      userfile_list = userfile.userfiles_accessible_by_user!(user, nil, nil, file_access_symbol)
+      userfile_list.compact.each do |subfile|
+        make_available(subfile, subfile.name)
+      end
+    end
+
+    true
+  end
+
+  # It adjusts the command-line of the descriptor.
+  # So that the token for the parent directory is replaced by the
+  # name of the parent directory we created in setup().
+  #
+  # From:
+  #
+  #   "command-line": "apptool [PRECOMPUTED_INPUT] [OUTPUT] stuff"
+  #
+  # To:
+  #
+  #   "command-line": "true [PRECOMPUTED_INPUT]; apptool #{parent_directory}/#{usefile.name} [OUTPUT] stuff"
+  #
+  # The reason a dummy true statement is prefixed at the beginning of the command
+  # is so that bosh won't complain if it can't find the token [PRECOMPUTED_INPUT] anywhere
+  # in the string.
+  def descriptor_for_cluster_commands
+    descriptor = super.dup
+
+    command    = descriptor.command_line
+
+    boutiques_dir_maker_list = descriptor.custom_module_info('BoutiquesDirMaker')
+    boutiques_dir_maker_list.each do |inputid,dirname|
+      input         = descriptor.input_by_id(inputid)
+      token         = input.value_key # e.g. '[PRECOMPUTED_INPUT]'
+      userfile_id   = invoke_params[inputid]
+      userfile_name = Userfile.find(userfile_id).name
+
+      # Make the substitution
+      command = command.sub(token, "#{dirname}/#{userfile_name}")
+
+      # In order to prevent bosh from complaining if the value-key is no longer found
+      # anywhere in the command-line, we re-insert a dummy no-op bash statement at the
+      # beginning of the command with at least one use of that value-key. It will look
+      # like e.g.
+      #
+      #   "true [PRECOMPUTED_INPUT] ; real command here"
+      #
+      # In bash, the 'true' statement doesn't do anything and ignores all arguments.
+      if ! command.include? token
+        command = "true #{token} ; #{command}"
```

Comment:
See my comment above, when adjusting the command line properly, you don't need any of these hacks here.